Online Boutique's Helm chart

mathieu-benoit opened the pull request:
Online Boutique's Helm chart - https://github.com/GoogleCloudPlatform/microservices-demo/issues/1319.


You can render the hydrated/generated manifests of this Helm chart from its associated folder:
```
cd helm-chart
helm template .
```
Or via the manually pushed remote Helm chart:
```
helm template oci://us-docker.pkg.dev/online-boutique-ci/charts/onlineboutique
```

You can do the same with both for installing it in your cluster:
```
cd helm-chart
helm upgrade onlineboutique . --install
```
Or:
```
helm upgrade onlineboutique oci://us-docker.pkg.dev/online-boutique-ci/charts/onlineboutique --install
```

Action items:
- [X] Create the Helm chart Online Boutique
- [X] Manually push the Helm chart in Artifact Registry
  - `cd helm-chart` && `helm package .` && `helm push onlineboutique-0.4.2.tgz oci://us-docker.pkg.dev/online-boutique-ci/charts`
- [X] Tests
- [x] Reviews
- [x] Script to push the Helm chart in GAR in the release process

mathieu-benoit commented on kustomize/components/google-cloud-operations/kustomization.yaml:
```diff
@@ -17,23 +17,8 @@ kind: Component
-            - name: ENABLE_TRACING
-              value: "1"
-# checkoutservice - stats, tracing, profiler
+# adservice - not yet implemented
```
Because I implemented the Cloud Ops options in the Helm chart, I took the initiative to update the associated Kustomize overlay to be accurate with what we have in the repo and not misleading.

mathieu-benoit commented on kubernetes-manifests/checkoutservice.yaml:
```diff
@@ -64,8 +64,6 @@ spec:
             value: "currencyservice:7000"
           - name: CART_SERVICE_ADDR
             value: "cartservice:7070"
-          - name: DISABLE_PROFILER
```
Because I implemented the Cloud Ops options in the Helm chart, I took the initiative to update the associated Kubernetes manifest to be accurate with what we have in the repo and not misleading. It's actually `ENABLE_PROFILER` for `checkoutservice` now.

NimJay commented on helm-chart/Chart.yaml:
```diff
@@ -0,0 +1,24 @@
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "v0.4.2"
```
**Issue:**
Please add automation to `/hack/make-release.sh` such that:
1. the versions inside this `Chart.yaml` file are updated.
2. the file is committed. 
Ideally, we would simplify the release process (such that the automation is no longer required), but that would require an audit of docs/demos using Online Boutique.

mathieu-benoit replied:
Done.

NimJay commented on helm-chart/README.md:
```diff
@@ -2,6 +2,8 @@
 
 If you'd like to deploy Online Boutique via its Helm chart, you could leverage the following instructions.
 
+**Warning:** Online Boutique's Helm chart is currently experimental. If you have feedback or run into issues, let us know inside [GitHub Issue #1319](https://github.com/GoogleCloudPlatform/microservices-demo/issues/1319) or by creating a [new GitHub Issue](https://github.com/GoogleCloudPlatform/microservices-demo/issues/new/choose).
```
**My rationale for this change:**
I've added a warning about these Helm charts being in experimental mode. We can remove this note once:
1. We have at least 1 review from another Helm expert from our team.
2. We have figured out how Helm fits into our current release process.
3. We are confident with the amount of documentation around these Helm charts (for both Online Boutique maintainers and users).

NimJay commented on .github/workflows/helm-chart-ci.yaml:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2020 Google LLC
+# Copyright 2022 Google LLC
```
**Rationale for this change:**
This is a nitpick.
I've updated the Apache license header years to 2022 for new files.

NimJay commented on helm-chart/templates/NOTES.txt:
```diff
@@ -0,0 +1,7 @@
+           You can watch the status of by running 'kubectl get --namespace {{ .Release.Namespace }} svc -w {{ .Values.frontend.name }}-external'
+  export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ .Values.frontend.name }}-external --template "{{"{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}"}}")
+  echo http://$SERVICE_IP
+{{- end }}
```
Praise:
This user experience is cool. I love that we can apply it conditionally (e.g., based on `.Values.frontend.externalService`).

NimJay commented on helm-chart/templates/frontend.yaml:
```diff
@@ -0,0 +1,271 @@
+  ports:
+  - name: http
+    port: 80
+    targetPort: 8080
```
**Question:**
This discussion is out-of-scope for this pull-request.
It looks like we will always have a `ClusterIP` `Service` called `frontend`.
It also looks like each microservice has a `Service` of type `ClusterIP`.
Do you think we should introduce a Kustomize Component for this (for non-Helm deployments of Online Boutique)? What is the main value of this?

NimJay commented on helm-chart/values.yaml:
```diff
@@ -0,0 +1,114 @@
+
+sidecars:
+  # Specifies if the Sidecars are created or not. If true, one fine granular Sidecar per app is created.
+  create: false
```
**Praise:** I appreciate the consist user experience here (`foo.create `).

NimJay reviewed (changes requested):
Thanks again for doing this, Mathieu!
I've made some small commits myself to the `mathieu-benoit/helm-chart` branch to reduce back-and-forth.

## Failed Testing

I've tested `helm template .` It worked. 👍 
However, when I ran `helm upgrade onlineboutique . --install`, I got the following error:
```
Error: UPGRADE FAILED: failed to create resource: Deployment in version "v1" cannot be handled as a Deployment: json: cannot unmarshal bool into Go struct field EnvVar.spec.template.spec.containers.env.value of type string
```
I got the same error for `helm upgrade onlineboutique oci://us-docker.pkg.dev/online-boutique-ci/charts/onlineboutique --install`. 🤔
Did it work for you?

Running `helm version` gives me:
```
version.BuildInfo{Version:"v3.9.3", GitCommit:"414ff28d4029ae8c8b05d62aa06c7fe3dee2bc58", GitTreeState:"clean", GoVersion:"go1.17.13"}
```
I'm doing all this on Cloud Shell with my `kubectx` set to a GKE cluster.

## Other Concerns
I've left some comments.
My main concerns are:
1. `hack/make-release.sh`
2. microservice name variables (e.g., `.Values.cartservice.name`)

mathieu-benoit commented:
Thanks for your review @NimJay, will tackle your request/feedback later.

FYI: I just fixed the issue you faced, `us-docker.pkg.dev/online-boutique-ci/charts/onlineboutique` is also updated accordingly, sorry about that. That should be good to go now.

NimJay commented on hack/make-helm-chart.sh:
```diff
@@ -0,0 +1,33 @@
+helm package .
+helm push onlineboutique-$TAG.tgz oci://$HELM_CHART_REPO
+
+log "Successfully built and pushed the Helm chart."
```
**Praise:** Nice! We can remove the `v` prefix with `{TAG:1}`.

NimJay commented on hack/make-helm-chart.sh:
```diff
@@ -0,0 +1,33 @@
+sed -i "s/^appVersion:.*/appVersion: \"${TAG}\"/" Chart.yaml
+sed -i "s/^version:.*/version: ${TAG:1}/" Chart.yaml
+helm package .
+helm push onlineboutique-$TAG.tgz oci://$HELM_CHART_REPO
```
**Praise:**
Thanks so much for adding this automation to `/hack` and doing it in a clean manner (e.g., introducing a separate `make-helm-chart.sh` file).

NimJay reviewed (approved):
Thanks for the quick turn-around, @mathieu-benoit!
My tests worked!
1. I ran: `helm upgrade onlineboutique . --install`
1. I saw:
    ```
    Release "onlineboutique" has been upgraded. Happy Helming!
    NAME: onlineboutique
    LAST DEPLOYED: Thu Dec  8 14:00:28 2022
    NAMESPACE: default
    STATUS: deployed
    REVISION: 4
    TEST SUITE: None
    NOTES:
    Get the application URL by running these commands:
      NOTE: It may take a few minutes for the LoadBalancer IP to be available.
      You can watch the status of by running 'kubectl get --namespace default svc -w frontend-external'
      export SERVICE_IP=$(kubectl get svc --namespace default frontend-external --template "{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}")
      echo http://$SERVICE_IP
    ```
1. I ran the 3 commands from the output above. All 3 worked! 👍 
1. I visited the URL for Online Boutique. Worked!
1. `helm upgrade onlineboutique . --install --set frontend.cymbalBranding=true`
1. I revisited the URL for Online Boutique and saw the Cymbal logo, and successfully checked out my cart.

![Screenshot 2022-12-08 at 2 15 30 PM](https://user-images.githubusercontent.com/10292865/206546809-27c42341-c5b3-43ed-89fa-447466ee2cde.png)

That UX was amazing! 🤯
